lib/spack/spack/spec.py

```diff
@@ -3700,6 +3702,18 @@ def __getitem__(self, name: str):
 
         return child
 
+    def __len__(self):
+        if not self.concrete:
+            raise spack.error.SpecError(f"Cannot get length of abstract spec: {self}")
+
+        if not self._length:
+            self._length = 1 + sum(len(dep) for dep in self.dependencies())
+        return self._length
+
+    def __bool__(self):
+        # Need to define this so __len__ isn't used by default
+        return True
```

Comment:
I see this PR changed a lot since last week. Was it necessary to use dunder methods with such an unusual implementation? 

Dunder methods may enter in different protocols, and given specs are in no way similar to any built-in container, this may cause subtle bugs which a named method would avoid. I think the comment itself indicates that we are violating one of the expectations on the behavior of `__len__` and `__bool__`.

Reply:
I'd *personally* rather see these moved out of `Spec` and into the code that needs them, as I don't think (especially with __bool__ the way it is) that they're general (or intuitive) enough to stay in `Spec`.

Reply:
See https://github.com/spack/spack/pull/47585